metrics.nim

```
@@ -759,59 +806,72 @@ when defined(metrics) and defined(linux):
+      result[@[]] = @[
+        Metric(
+          name: "process_virtual_memory_bytes", # Virtual memory size in bytes.
+          value: selfStat[20].parseFloat(),
+          timestamp: timestamp,
+        ),
+        Metric(
+          name: "process_resident_memory_bytes", # Resident memory size in bytes.
+          value: selfStat[21].parseFloat() * pagesize,
+          timestamp: timestamp,
+        ),
+        Metric(
+          name: "process_start_time_seconds", # Start time of the process since unix epoch in seconds.
+          value: selfStat[19].parseFloat() / ticks + btime,
+          timestamp: timestamp,
+        ),
+        Metric(
+          name: "process_cpu_seconds_total", # Total user and system CPU time spent in seconds.
+          value: (selfStat[11].parseFloat() + selfStat[12].parseFloat()) / ticks,
+          timestamp: timestamp,
+        ),
+      ]
+
+      for line in lines("/proc/self/limits"):
+        if line.startsWith("Max open files"):
+          result[@[]].add(
+            Metric(
+              name: "process_max_fds", # Maximum number of open file descriptors.
+              value: line.split(whitespaceRegex)[3].parseFloat(), # a simple `split()` does not combine adjacent whitespace
+              timestamp: timestamp,
+            )
+          )
+          break
 
-    for line in lines("/proc/self/limits"):
-      if line.startsWith("Max open files"):
+      result[@[]].add(
+        Metric(
+          name: "process_open_fds", # Number of open file descriptors.
+          value: toSeq(walkDir("/proc/self/fd")).len.float64,
+          timestamp: timestamp,
+        )
+      )
+    except CatchableError:
+      try:
         result[@[]].add(
           Metric(
-            name: "process_max_fds", # Maximum number of open file descriptors.
-            value: line.split(whitespaceRegex)[3].parseFloat(), # a simple `split()` does not combine adjacent whitespace
+            name: "metric_error",
```

The best way to hide an error I've seen so far. Nobody will add this metric name to Grafana, so no one will see it.

---

you're doing the same thing when `/proc` is not available, except without the metric - what's your suggestion? discard? crash application if `/proc` has issues?

---

"/proc" not being mounted is not a showstopper, nor is it something the API user needs to be informed about by this library.